File: BOJ/4307.cpp
#include <bits/stdc++.h>
using namespace std;

pair<int, int> Solve(int len, const vector<int>& ants) {
	int fast_time = 0; // 모든 개미에 대한 최단 시간 중 가장 큰 값
	int late_time = 0; // 모든 개미에 대한 최장 시간 중 가장 큰 값
	for (int i = 0; i < ants.size(); i++) { // 각각의 개미에 대해 최단, 최장시간을 구해준다.
		fast_time = max(fast_time, min(abs(len - ants[i]), ants[i]));
		late_time = max(late_time, max(abs(len - ants[i]), ants[i]));
	}

	return make_pair(fast_time, late_time);
}

int main() {
	int tc;
	cin >> tc; // test case 입력
	while (tc--) {
		int len, num_of_ants;
		cin >> len >> num_of_ants; // 막대의 길이와 개미의 수 입력

		vector<int> ants;
		for (int i = 0; i < num_of_ants; i++) { // 개미의 위치 입력
			int loc;
			cin >> loc;
			ants.push_back(loc);
		}
		pair<int, int> answer = Solve(len, ants);

		cout << answer.first << ' ' << answer.second << '\n';
	}

	return 0;
}

File: BOJ/python/4307.py
import sys
input = sys.stdin.readline

def solve(length, ants):
    fast_time = late_time = 0
    for i in range(len(ants)):
        fast_time = max(fast_time, min(abs(length - ants[i]), ants[i]))  # 각 개미의 최단 시간
        late_time = max(late_time, max(abs(length - ants[i]), ants[i]))  # 각 개미의 최장 시간

    return fast_time, late_time


tc = int(input())
for i in range(tc):
    length, num_of_ants = map(int, input().split())
    ants = [int(input()) for k in range(num_of_ants)]
    fast, last = solve(length, ants)
    print(fast, last)
